### PR TITLE
Add possibility to filter yaml keys

### PR DESCRIPTION
**Pull request by matthiashamacher**

### CODE REVIEW

**Comment by matthiashamacher** [comment]
We just now noticed that the fix:structure Command seems to be removing the filtered entries from the result file. I will have a look how we can fix that, but so long this should not be merged and released.

File: src/Traits/ArrayTrait.php [hunk 1]
```diff
@@ -13,15 +15,19 @@ trait ArrayTrait
...
      * @return array<mixed>
      */
-    protected function getFlatArray(array $array, string $delimiter, string $rootPrefix = ''): array
+    protected function getFlatArray(array $array, string $delimiter, string $rootPrefix = '', ?Filter $filter = null): array
     {
         $result = [];
 
         foreach ($array as $key => $value) {
+            if ($filter instanceof Filter && $filter->isKeyAllowed($key) === false) {
+                continue;
```
Comment (matthiashamacher) [inline src/Traits/ArrayTrait.php]: If we don't continue here but save the filtered keys into an additional array, we can remove them during loading and merging those two array together during saving/writing.

@boxblinkracer Would that be a feasible solution for the problem that the filtered keys are remove with the `fix:structure` command?

Comment (boxblinkracer) [reply src/Traits/ArrayTrait.php]: hi
yes thats absolutely fine, a really good idea
so is that now already working, can i merge it like this @matthiashamacher ?

Comment (matthiashamacher) [reply src/Traits/ArrayTrait.php]: No not right now. Wanted your input first so I haven't added it. Will add it later today and ping you so you can merge it.

Comment (matthiashamacher) [reply src/Traits/ArrayTrait.php]: Finally found to time to finalize this pr. Was a little bit more complicated than I thought but I think its an ok solution @boxblinkracer Please have a look.

Comment (boxblinkracer) [reply src/Traits/ArrayTrait.php]: hi @matthiashamacher 
awesome thank you, ill check it out later or this weekend :)
thanks for your work on that, means a lot

File: src/Traits/ArrayTrait.php [hunk 2]
```diff
@@ -30,6 +32,30 @@ protected function getFlatArray(array $array, string $delimiter, string $rootPre
...
+     *
+     * @return array<int, string[]>
+     */
+    protected function getFilteredResult(array $array, Filter $filter): array
+    {
+        $translations = [];
+        $filteredTranslations = [];
+
+        foreach ($array as $key => $value) {
+            if ($filter->isKeyAllowed($key, true) === true) {
```
Comment (boxblinkracer) [inline src/Traits/ArrayTrait.php:47]: the only thing that i dont really get is the partial allowing?
why do we have that?
can you also explain that in a comment.
if i want to filter an array, i dont see a point where i would need partial filtering

**Review by boxblinkracer — CHANGES_REQUESTED** [review]

File: src/Traits/ArrayTrait.php (hunk 2 above)
Comment (matthiashamacher) [reply src/Traits/ArrayTrait.php:47]: Before I add the comment I explain it here, maybe there is a better way for the whole solution.

Background is that filtering during the first run through of flattening the array was no good solution in my opinion. During implementation there were too many changes needed to be done to save the filtered results. Therefore I moved it to this new function. As this runs after the array is flattened we have an array like this:
```
$array = [
    'app.name' => 'test',
    'app.description' => 'test',
];
```
If we now want to filter out description, we cannot filter by an exact match as the string would is `app.description` therefore we need the partial match.

**Pull request closed without merging**